File: backend/src/services/SchedulingService.ts
import { Process, ProcessResult, SchedulingResult } from "../types/scheduling";
import { SystemInfoService } from "./SystemInfoService";

export class SchedulingService {
  private systemInfo = SystemInfoService.getSystemInfo();
  private cores = this.systemInfo.cores;

  private calculateAverageMetrics(
    results: ProcessResult[]
  ): SchedulingResult["averageMetrics"] {
    const totalProcesses = results.length;
    const totalWaitingTime = results.reduce((sum, p) => sum + p.waitingTime, 0);
    const totalResponseTime = results.reduce(
      (sum, p) => sum + p.responseTime,
      0
    );
    const totalTurnaroundTime = results.reduce(
      (sum, p) => sum + p.turnaroundTime,
      0
    );
    const maxCompletionTime = Math.max(...results.map((p) => p.completionTime));

    return {
      waitingTime: totalWaitingTime / totalProcesses,
      responseTime: totalResponseTime / totalProcesses,
      turnaroundTime: totalTurnaroundTime / totalProcesses,
      throughput: totalProcesses / maxCompletionTime,
      cpuUtilization:
        (totalTurnaroundTime / (this.cores * maxCompletionTime)) * 100,
    };
  }

  private createTimeline(
    results: ProcessResult[]
  ): SchedulingResult["timeline"] {
    const timeline: SchedulingResult["timeline"] = {};

    for (let i = 0; i < this.cores; i++) {
      timeline[i] = { processes: [] };
    }

    results.forEach((process) => {
      timeline[process.cpuCore].processes.push({
        id: process.id,
        startTime: process.startTime,
        endTime: process.endTime,
      });
    });

    return timeline;
  }

  private formatResult(
    algorithm: string,
    processResults: ProcessResult[]
  ): SchedulingResult {
    return {
      algorithm,
      systemInfo: {
        cores: this.systemInfo.cores,
        cpuModel: this.systemInfo.cpuModel,
        architecture: this.systemInfo.architecture,
      },
      averageMetrics: this.calculateAverageMetrics(processResults),
      processResults,
      timeline: this.createTimeline(processResults),
    };
  }

  public fcfs(processes: Process[]): SchedulingResult {
    const sortedProcesses = [...processes].sort(
      (a, b) => a.arrivalTime - b.arrivalTime
    );
    const results: ProcessResult[] = [];
    const coreTimes = new Array(this.cores).fill(0); // Track time for each core

    sortedProcesses.forEach((process) => {
      // Find the core with minimum current time
      const minCoreIndex = coreTimes.indexOf(Math.min(...coreTimes));
      const startTime = Math.max(coreTimes[minCoreIndex], process.arrivalTime);
      const endTime = startTime + process.burstTime;

      results.push({
        id: process.id,
        waitingTime: startTime - process.arrivalTime,
        responseTime: startTime - process.arrivalTime,
        completionTime: endTime,
        turnaroundTime: endTime - process.arrivalTime,
        cpuCore: minCoreIndex,
        startTime,
        endTime,
      });

      coreTimes[minCoreIndex] = endTime;
    });

    return this.formatResult("FCFS", results);
  }

  public sjf(processes: Process[]): SchedulingResult {
    const results: ProcessResult[] = [];
    const coreTimes = new Array(this.cores).fill(0);
    let remainingProcesses = [...processes];
    let currentTime = Math.min(...processes.map((p) => p.arrivalTime));

    while (remainingProcesses.length > 0) {
      const availableProcesses = remainingProcesses.filter(
        (p) => p.arrivalTime <= currentTime
      );

      if (availableProcesses.length === 0) {
        currentTime = Math.min(...remainingProcesses.map((p) => p.arrivalTime));
        continue;
      }

      // Find shortest job among available processes
      const shortestJob = availableProcesses.reduce((min, p) =>
        p.burstTime < min.burstTime ? p : min
      );

      // Find core with minimum current time
      const minCoreIndex = coreTimes.indexOf(Math.min(...coreTimes));
      const startTime = Math.max(coreTimes[minCoreIndex], currentTime);
      const endTime = startTime + shortestJob.burstTime;

      results.push({
        id: shortestJob.id,
        waitingTime: startTime - shortestJob.arrivalTime,
        responseTime: startTime - shortestJob.arrivalTime,
        completionTime: endTime,
        turnaroundTime: endTime - shortestJob.arrivalTime,
        cpuCore: minCoreIndex,
        startTime,
        endTime,
      });

      coreTimes[minCoreIndex] = endTime;
      currentTime = Math.min(...coreTimes);
      remainingProcesses = remainingProcesses.filter(
        (p) => p.id !== shortestJob.id
      );
    }

    return this.formatResult("SJF", results);
  }

  public priority(processes: Process[]): SchedulingResult {
    const results: ProcessResult[] = [];
    const coreTimes = new Array(this.cores).fill(0);
    let remainingProcesses = [...processes];
    let currentTime = Math.min(...processes.map((p) => p.arrivalTime));

    while (remainingProcesses.length > 0) {
      const availableProcesses = remainingProcesses.filter(
        (p) => p.arrivalTime <= currentTime
      );

      if (availableProcesses.length === 0) {
        currentTime = Math.min(...remainingProcesses.map((p) => p.arrivalTime));
        continue;
      }

      // Find highest priority process (lower number = higher priority)
      const highestPriority = availableProcesses.reduce((min, p) =>
        p.priority < min.priority ? p : min
      );

      // Find core with minimum current time
      const minCoreIndex = coreTimes.indexOf(Math.min(...coreTimes));
      const startTime = Math.max(coreTimes[minCoreIndex], currentTime);
      const endTime = startTime + highestPriority.burstTime;

      results.push({
        id: highestPriority.id,
        waitingTime: startTime - highestPriority.arrivalTime,
        responseTime: startTime - highestPriority.arrivalTime,
        completionTime: endTime,
        turnaroundTime: endTime - highestPriority.arrivalTime,
        cpuCore: minCoreIndex,
        startTime,
        endTime,
      });

      coreTimes[minCoreIndex] = endTime;
      currentTime = Math.min(...coreTimes);
      remainingProcesses = remainingProcesses.filter(
        (p) => p.id !== highestPriority.id
      );
    }

    return this.formatResult("Priority", results);
  }

  public roundRobin(
    processes: Process[],
    timeQuantum: number = 2
  ): SchedulingResult {
    const results: ProcessResult[] = [];
    const coreTimes = new Array(this.cores).fill(0);
    let remainingProcesses = processes.map((p) => ({
      ...p,
      remainingTime: p.burstTime,
      firstResponse: -1,
    }));
    let currentTime = Math.min(...processes.map((p) => p.arrivalTime));
    const processStartTimes = new Map<number, number>();

    while (remainingProcesses.length > 0) {
      const availableProcesses = remainingProcesses.filter(
        (p) => p.arrivalTime <= currentTime
      );

      if (availableProcesses.length === 0) {
        currentTime = Math.min(...remainingProcesses.map((p) => p.arrivalTime));
        continue;
      }

      // Process each available process for the time quantum
      for (const process of availableProcesses) {
        const minCoreIndex = coreTimes.indexOf(Math.min(...coreTimes));
        const startTime = Math.max(coreTimes[minCoreIndex], currentTime);
        const executionTime = Math.min(timeQuantum, process.remainingTime);
        const endTime = startTime + executionTime;

        // Track first response time
        if (process.firstResponse === -1) {
          process.firstResponse = startTime;
        }

        // If process is completing
        if (process.remainingTime <= timeQuantum) {
          results.push({
            id: process.id,
            waitingTime:
              startTime -
              process.arrivalTime -
              (process.burstTime - process.remainingTime),
            responseTime: process.firstResponse - process.arrivalTime,
            completionTime: endTime,
            turnaroundTime: endTime - process.arrivalTime,
            cpuCore: minCoreIndex,
            startTime,
            endTime,
          });
        }

        coreTimes[minCoreIndex] = endTime;
        process.remainingTime -= executionTime;
      }

      currentTime = Math.min(...coreTimes);
      remainingProcesses = remainingProcesses.filter(
        (p) => p.remainingTime > 0
      );
    }

    return this.formatResult("Round Robin", results);
  }
}

File: backend/services/scheduling_service.py
from .system_info_service import SystemInfoService

class SchedulingService:
    def __init__(self):
        self.system_info = SystemInfoService.get_system_info()
        self.cores = self.system_info['cores']  # Using physical cores for accurate simulation

    def _calculate_average_metrics(self, results):
        if not results:
            return {
                'waiting_time': 0,
                'response_time': 0,
                'turnaround_time': 0,
                'throughput': 0,
                'cpu_utilization': 0
            }

        total_processes = len(results)
        total_waiting_time = sum(p['waiting_time'] for p in results)
        total_response_time = sum(p['response_time'] for p in results)
        total_turnaround_time = sum(p['turnaround_time'] for p in results)
        max_completion_time = max(p['completion_time'] for p in results)

        return {
            'waiting_time': total_waiting_time / total_processes,
            'response_time': total_response_time / total_processes,
            'turnaround_time': total_turnaround_time / total_processes,
            'throughput': total_processes / max_completion_time if max_completion_time > 0 else 0,
            'cpu_utilization': (total_turnaround_time / (self.cores * max_completion_time) * 100) if max_completion_time > 0 else 0
        }

    def _create_timeline(self, results):
        timeline = {str(i): {'processes': []} for i in range(self.cores)}
        
        for process in results:
            core = str(process['cpu_core'])
            timeline[core]['processes'].append({
                'id': process['id'],
                'start_time': process['start_time'],
                'end_time': process['end_time']
            })
        
        return timeline

    def _format_result(self, algorithm, process_results):
        return {
            'algorithm': algorithm,
            'system_info': {
                'cores': self.system_info['cores'],
                'cpu_model': self.system_info['cpu_model'],
                'architecture': self.system_info['architecture']
            },
            'average_metrics': self._calculate_average_metrics(process_results),
            'process_results': process_results,
            'timeline': self._create_timeline(process_results)
        }

    def fcfs(self, processes):
        sorted_processes = sorted(processes, key=lambda x: x['arrival_time'])
        results = []
        core_times = [0] * self.cores

        for process in sorted_processes:
            # Find core with minimum current time
            min_core_index = core_times.index(min(core_times))
            start_time = max(core_times[min_core_index], process['arrival_time'])
            end_time = start_time + process['burst_time']

            results.append({
                'id': process['id'],
                'waiting_time': start_time - process['arrival_time'],
                'response_time': start_time - process['arrival_time'],
                'completion_time': end_time,
                'turnaround_time': end_time - process['arrival_time'],
                'cpu_core': min_core_index,
                'start_time': start_time,
                'end_time': end_time
            })

            core_times[min_core_index] = end_time

        return self._format_result('FCFS', results)

    def sjf(self, processes):
        results = []
        core_times = [0] * self.cores
        remaining_processes = processes.copy()
        current_time = min(p['arrival_time'] for p in processes)

        while remaining_processes:
            available_processes = [
                p for p in remaining_processes
                if p['arrival_time'] <= current_time
            ]

            if not available_processes:
                current_time = min(p['arrival_time'] for p in remaining_processes)
                continue

            # Find shortest job
            shortest_job = min(available_processes, key=lambda x: x['burst_time'])
            min_core_index = core_times.index(min(core_times))
            start_time = max(core_times[min_core_index], current_time)
            end_time = start_time + shortest_job['burst_time']

            results.append({
                'id': shortest_job['id'],
                'waiting_time': start_time - shortest_job['arrival_time'],
                'response_time': start_time - shortest_job['arrival_time'],
                'completion_time': end_time,
                'turnaround_time': end_time - shortest_job['arrival_time'],
                'cpu_core': min_core_index,
                'start_time': start_time,
                'end_time': end_time
            })

            core_times[min_core_index] = end_time
            current_time = min(core_times)
            remaining_processes.remove(shortest_job)

        return self._format_result('SJF', results)

    def priority(self, processes):
        results = []
        core_times = [0] * self.cores
        remaining_processes = processes.copy()
        current_time = min(p['arrival_time'] for p in processes)

        while remaining_processes:
            available_processes = [
                p for p in remaining_processes
                if p['arrival_time'] <= current_time
            ]

            if not available_processes:
                current_time = min(p['arrival_time'] for p in remaining_processes)
                continue

            # Find highest priority process (lower number = higher priority)
            highest_priority = min(available_processes, key=lambda x: x['priority'])
            min_core_index = core_times.index(min(core_times))
            start_time = max(core_times[min_core_index], current_time)
            end_time = start_time + highest_priority['burst_time']

            results.append({
                'id': highest_priority['id'],
                'waiting_time': start_time - highest_priority['arrival_time'],
                'response_time': start_time - highest_priority['arrival_time'],
                'completion_time': end_time,
                'turnaround_time': end_time - highest_priority['arrival_time'],
                'cpu_core': min_core_index,
                'start_time': start_time,
                'end_time': end_time
            })

            core_times[min_core_index] = end_time
            current_time = min(core_times)
            remaining_processes.remove(highest_priority)

        return self._format_result('Priority', results)

    def round_robin(self, processes, time_quantum=2):
        results = []
        core_times = [0] * self.cores
        remaining_processes = [{
            **p,
            'remaining_time': p['burst_time'],
            'first_response': -1
        } for p in processes]
        current_time = min(p['arrival_time'] for p in processes)

        while remaining_processes:
            available_processes = [
                p for p in remaining_processes
                if p['arrival_time'] <= current_time
            ]

            if not available_processes:
                current_time = min(p['arrival_time'] for p in remaining_processes)
                continue

            # Process each available process for the time quantum
            for process in available_processes:
                min_core_index = core_times.index(min(core_times))
                start_time = max(core_times[min_core_index], current_time)
                execution_time = min(time_quantum, process['remaining_time'])
                end_time = start_time + execution_time

                # Track first response time
                if process['first_response'] == -1:
                    process['first_response'] = start_time

                # If process is completing
                if process['remaining_time'] <= time_quantum:
                    results.append({
                        'id': process['id'],
                        'waiting_time': start_time - process['arrival_time'] - (process['burst_time'] - process['remaining_time']),
                        'response_time': process['first_response'] - process['arrival_time'],
                        'completion_time': end_time,
                        'turnaround_time': end_time - process['arrival_time'],
                        'cpu_core': min_core_index,
                        'start_time': start_time,
                        'end_time': end_time
                    })

                core_times[min_core_index] = end_time
                process['remaining_time'] -= execution_time

            current_time = min(core_times)
            remaining_processes = [p for p in remaining_processes if p['remaining_time'] > 0]

        return self._format_result('Round Robin', results) 